semantic/semantic.fbs

```diff
@@ -0,0 +1,354 @@
+}
+
+table SourceLocation {
+    file:string;
+    start:Position;
+    end:Position;
+    source:string;
+}
+
+struct Position {
+    line:int;
+    column:int;
+}
+
+// Statements
+
+// Unions cannot be vector elements,
+// so wrap the union in a table.
+table WrappedStatement {
+    statement:Statement;
+}
+
+table OptionStatement {
+    loc:SourceLocation;
+    assignment:Assignment;
+}
+
+table BuiltinStatement {
+    loc:SourceLocation; 
+    id:Identifier; 
+}
+
+table TestStatement {
+    loc:SourceLocation; 
+    assignment:NativeVariableAssignment;
+}
+
+table ExpressionStatement {
+    loc:SourceLocation; 
+    expression:Expression; 
+}
+
+table ReturnStatement {
+    loc:SourceLocation; 
+    argument:Expression; 
+}
+
+table NativeVariableAssignment {
+    loc:SourceLocation;
+    identifier:Identifier; 
+    init_:Expression; 
+}
+
+table MemberAssignment {
+    loc:SourceLocation; 
+    member:MemberExpression;
+    init_:Expression; 
+}
+
+table Extern {
+    loc:SourceLocation; 
+    assignments:[ExternalVariableAssignment];
+    block:ExternBlock; 
+}
+
+table ExternalVariableAssignment {
+    loc:SourceLocation; 
+    identifier:Identifier;
+}
+
+table ExternBlock {
+    loc:SourceLocation;
+    node:Block;
+}
+
+// Expressions
+
+// Unions cannot be vector elements,
+// so wrap the union in a table.
+table WrappedExpression {
+    expression:Expression;
+}
+
+table StringExpression {
+    loc:SourceLocation;
+    parts:[StringExpressionPart];
+}
+
+table StringExpressionPart {
+    loc:SourceLocation;
+    // Following two fields are mutually exclusive:
+    text_value:string; // for TextPart
+    interpolated_expression:Expression; // for InterpolatedPart
+}
+
+table ArrayExpression {
+    loc:SourceLocation;
+    elements:[WrappedExpression];
+}
+
+table FunctionExpression {
+    loc:SourceLocation;
+    defaults:ObjectExpression;
+    Block:FunctionBlock;
+}
+
+table FunctionBlock {
+    loc:SourceLocation;
+    parameters:FunctionParameters;
+    body:Block;
+}
+
+table FunctionParameters {
+    loc:SourceLocation;
+    list:[FunctionParameter];
+    pipe:Identifier;
+}
+
+table FunctionParameter {
+    loc:SourceLocation;
+    key:Identifier;
+}
+
+enum Operator : byte {
+  MultiplicationOperator,
+  DivisionOperator,
+  ModuloOperator,
+  PowerOperator,
+  AdditionOperator,
+  SubtractionOperator,
+  LessThanEqualOperator,
+  LessThanOperator,
+  GreaterThanEqualOperator,
+  GreaterThanOperator,
+  StartsWithOperator,
+  InOperator,
+  NotOperator,
+  ExistsOperator,
+  NotEmptyOperator,
+  EmptyOperator,
+  EqualOperator,
+  NotEqualOperator,
+  RegexpMatchOperator,
+  NotRegexpMatchOperator,
+  InvalidOperator,
+}
+
+enum LogicalOperator : byte {
+  AndOperator,
+  OrOperator,
+}
+
+table BinaryExpression {
+    loc:SourceLocation;
+    operator:Operator; 
+    left: Expression;
+    right: Expression;
+}
+
+table CallExpression {
+    loc:SourceLocation;
+    callee:Expression;
+    arguments:[ObjectExpression];
```

Comment:
I made the AST FlatBuffers equivalent of this be a single `ObjectExpression`; it might be good to do the same here.

In the Go semantic graph, it's just a single `ObjectExpression` as well:
https://github.com/influxdata/flux/blob/master/semantic/graph.go

However in Rust, it's this:
https://github.com/influxdata/flux/blob/master/libflux/src/semantic/nodes.rs#L684